addon/-private/properties/collection/main.js

```diff
@@ -0,0 +1,140 @@
+/* global Symbol */
+import Ember from 'ember';
+import { buildSelector, assign } from '../../helpers';
+import { create } from '../../create';
+import { count } from '../count';
+import Ceibo from 'ceibo';
+
+import { throwBetterError } from '../../better-errors';
+
+export class Collection {
+  constructor(definition, parent, key) {
+    this.definition = definition;
+    this.parent = parent;
+    this.key = key;
+
+    this._itemCounter = create({
+      count: count(definition.scope, {
+        resetScope: definition.resetScope,
+        testContainer: definition.testContainer
+      })
+    }, { parent });
+
+    this._items = [];
+  }
+
+  objectAt(index) {
+    if (typeof this._items[index] === 'undefined') {
+      let { definition, parent, key } = this;
+      let scope = buildSelector({}, definition.scope, { at: index });
+
+      let finalizedDefinition = assign({}, definition);
+
+      finalizedDefinition.scope = scope;
+
+      let tree = create(finalizedDefinition, { parent });
+
+      // Change the key of the root node
+      Ceibo.meta(tree).key = `${key}[${index}]`;
+
+      this._items[index] = tree;
+    }
+
+    return this._items[index];
+  }
+
+  get length() {
+    return this._itemCounter.count;
+  }
+
+  toArray() {
+    let { length } = this;
+
+    let array = Ember.A();
+
+    for (let i = 0; i < length; i++) {
+      array.push(this.objectAt(i));
+    }
+
+    return array;
+  }
+
+  map(...args) {
+    return this.toArray().map(...args);
+  }
+
+  mapBy(...args) {
+    return this.toArray().mapBy(...args);
+  }
+
+  forEach(...args) {
+    return this.toArray().forEach(...args);
+  }
+
+  findOne(query) {
```

Comment:
Any reasons not to call it `find()`?

Reply:
`find` is a native array method and only accepts a function. `findOne` can receive a general query object, matching the signature of lodash's `find` method. It also asserts that _exactly_ one match exists, whereas `find` will find the _first_ instance, but also allows other instances to exist in the array. The difference in behavior is big enough we decided to rename the function to avoid confusion.

Reply:
> It also asserts that exactly one match exists

I believe this should be a responsibility of the test framework:
```js
  const visibleItems = page.items.findAll(i => i.isVisible);
  assert.equal(visibleItems.length, 1); // the user also can specify assertion message specific to the testing context
```

Reply:
> findOne can receive a general query object

I think understand the reason behind adding a declarative query objects feature. However I think it's quite complicated area to be done properly. I mean without any specific DSL it has only strict equals. We can't use `less then`, `greater then`, `contains`. Also it's not obvious how to deal with nested objects:
```js
page.items.findOne({
  isVisible: true,
  input: {
    value: 'test' // is this kind of comparison supported?
  }
})
``` 

And the last, by comparing function vs query object I can't see big benefits in using the former:
```js
page.items.findAll(i => i.isVisible)
page.items.findAll({ isVisible: true })
```

Reply:
A better example for the benefits of the query object would be with string values, or multiple values:

```
page.items.findAll(item => item.name === 'S&P 500 Stock Index' && item.isVisible);

page.items.findAll({ 
  name: 'S&P 500 Stock Index',
  isVisible: true
});
```

Really this is just a way to provide shorthand. The current API proposed in `findAll` and `findOne` was proposed in #345, originally I suggested adding the `findBy` method but @magistrula suggested using query objects instead since they are more flexible.

As for the assertion about only one result existing, I personally disagree. Relying on implicit behavior is the reason tests become brittle - for instance, relying on index rather than content of an item in a collection. If a user thinks they are selecting the _only_ instance of an item, but they are really selecting the _first_ instance, the resulting tests only have more chance of being brittle and breaking due to unrelated changes, especially in a way that is hard to debug. However, if we enforce this assertion then most users will get the behavior they want (find the one item they were looking for) and the rest can default to using `findAll` instead.

I'm also open to switching back to `filter` and `filterBy` and implementing `find` and `findBy` as well, but I think this is an opportunity to add more layers of safety that help prevent users from making brittle tests.

Reply:
It's also worth noting, we can do either strategy right now and implement the other later on. So we could stick to native array methods for now and consider adding `findOne` and `findAll` later on, or vice-versa. I'd definitely prefer that if we can get this merged since the other benefits are pretty massive.

Reply:
I'm not sure about the right path forward. Maybe @magistrula @san650 can help here?

There are several interesting features like query objects and `findOne` but I feel like they are out of scope here. I'd like to ship new collections in the slimmest possible form and iterate on new features in a separate PRs.

Reply:
If I cheat a little bit it I can't see benefits again 😆 :
```js
page.items.findAll(i => 
  i.name === 'S&P 500 Stock Index'
  && i.isVisible
);

page.items.findAll({ 
  name: 'S&P 500 Stock Index',
  isVisible: true
});
```

> since they are more flexible

I don't think object queries are flexible in its current form. I've tried to explain why [here](using query objects instead since they are more flexible).

Actually I would rather deprecate `filterBy` cause it seems like a weak api.

---

Regarding `findOne` you made me doubt @pzuraq. It seems more reasonable to me now.

Reply:
Seems like there's enough discussion of `findOne` and `findAll` here that I think it would make sense to simplify this PR by pulling those into a separate PR, but will that mean that we need to decide whether to provide `filter` and `filterBy` in this PR? Or can we leave those out as well and decide about them when we discuss `findOne` and `findAll`? Do we think that `filter`, `filterBy`, `findOne`, and `findAll` could all live alongside each other in the new API?